python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,1800 @@
+    ):
+        """
+        Tests Pub/Sub with two publishing clients using the same channel name.
+        One client uses pattern subscription, the other uses exact.
+        The clients publishes messages to each other, and to thyself.
+
+        This test verifies that two separate clients can correctly publish to and handle subscriptions
+        for exact and pattern channels with the same name. It covers the following scenarios:
+        - Subscribing to an exact channel and verifying message reception.
+        - Subscribing to a pattern channel and verifying message reception.
+        - Ensuring that messages are correctly published and received using different retrieval methods (async, sync, callback).
+        - Verifying that no messages are left unread.
+        - Properly unsubscribing from all channels to avoid interference with other tests.
+        """
+        CHANNEL_NAME = "same-channel-name" + get_random_string(3)
+        MESSAGE_BASIC = get_random_string(10)
+        MESSAGE_PATTERN = get_random_string(7)
+        publish_response = 2 if cluster_mode else OK
+        callback, context, context2 = None, None, None
+        callback_messages_basic: List[
+            Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg]
+        ] = []
+        callback_messages_pattern: List[
+            Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg]
+        ] = []
+
+        if method == MethodTesting.Callback:
+            callback = new_message
+            context = callback_messages_basic
+            context2 = callback_messages_pattern
+
+        # Setup pub/sub for exact channel
+        pub_sub_basic = create_pubsub_subscription(
+            cluster_mode,
+            {PubSubChannelModes.Exact: {CHANNEL_NAME}},
+            callback=callback,
+            context=context,
+        )
+        # Setup pub/sub for pattern channels
+        pub_sub_pattern = create_pubsub_subscription(
+            cluster_mode,
+            {PubSubChannelModes.Pattern: {CHANNEL_NAME}},
+            callback=callback,
+            context=context2,
+        )
+
+        client_pattern, client_basic = await setup_clients(
+            request, cluster_mode, pub_sub_basic, pub_sub_pattern
+        )
+
+        # Publish messages to each channel - both clients publishing
+        assert (
+            await client_pattern.publish(MESSAGE_BASIC, CHANNEL_NAME)
+            == publish_response
+        )
+        assert (
+            await client_basic.publish(MESSAGE_PATTERN, CHANNEL_NAME)
+            == publish_response
+        )
+
+        # allow the message to propagate
+        await asyncio.sleep(1)
+
+        # Verify message for basic pub/sub
+        for i in range(2):
+            pubsub_msg_basic = await get_message_by_method(
+                method, client_basic, callback_messages_basic, i
+            )
+            msg = MESSAGE_BASIC if i < 1 else MESSAGE_PATTERN
```

Comment:
this is wrong to assume the ordering of publish will be preserved upon the receptions. Consider the following example in standalone mode:
1. client_a connected to a primary
2. client_b connected to a primary and subscribed to channel "test"
3. client_c connected to a far replica, and publishes "message_from_c" to test channel. Replica returns OK on the publish request and starts to replicate the publish to the primary
4. client_a publishes the same "message_from_b" to test channel.
5. client_b immediately receives the message since he is on the same node that received publish from client_a